### PR TITLE
379 cart discount reference replacement and reference resolution

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
This PR addresses adding reference replacement for custom type references on cart discount sync. 

#### Description
For now reference resolution/replacement is only implemented for custom type references. References under GiftLineItem is **not** currently supported by this PR (and this release)

#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [x] Add Release Notes entry.

#### Hints for Review

The main implementation is in : 
- `CartDiscountReferenceReplacementUtils.java`
- `CartDiscountReferenceResolver.java` 
- `CartDiscountSync.java`

Integration Tests are in: 
- `CartDiscountITUtils.java`
- `externalsource/cartdiscounts/CartDiscountSyncIT.java`
- `ctpprojectsource/cartdiscounts/CartDiscountSyncIT.java`

Unit tests are in:
- `CartDiscountSyncTest.java`
- `CartDiscountReferenceResolverTest.java`
- `CartDiscountReferenceReplacementUtilsTest.java`

### CODE REVIEW

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/cartdiscounts/CartDiscountSyncIT.java [hunk 1]
```diff
@@ -84,26 +99,25 @@ static void tearDown() {
...
+        final List<Throwable> exceptions = new ArrayList<>();
+        final List<UpdateAction<CartDiscount>> updateActionsList = new ArrayList<>();
+
         final CartDiscountSyncOptions cartDiscountSyncOptions = CartDiscountSyncOptionsBuilder
             .of(CTP_TARGET_CLIENT)
+            .errorCallback((error, throwable) -> {
+                errorMessages.add(error);
+                exceptions.add(throwable);
+            })
+            .beforeUpdateCallback((updateActions, newCartDiscount, oldCartDiscount) -> {
```
Comment (ahmetoz) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/cartdiscounts/CartDiscountSyncIT.java]: it's good that you added update action checks to all tests 👍

File: src/main/java/com/commercetools/sync/cartdiscounts/helpers/CartDiscountReferenceResolver.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public CartDiscountReferenceResolver(
+        @Nonnull final CartDiscountSyncOptions options,
+        @Nonnull final TypeService typeService) {
+        super(options, typeService);
+    }
+
+    /**
+     * Given a {@link CartDiscountDraft} this method attempts to resolve the custom type reference to
+     * return a {@link CompletionStage} which contains a new instance of the draft with the resolved
+     * reference. If the {@link CartDiscountDraft} has a {@link GiftLineItemCartDiscountValue} value,
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/cartdiscounts/helpers/CartDiscountReferenceResolver.java]: > If the {@link CartDiscountDraft} has a {@link GiftLineItemCartDiscountValue} value,
     * the resourceIdentifiers (namely: product, supplyChannel and distributionChannel) are validated that they have
     * non-blank (empty/null) keys, if they are provided.

Where is this check? And why do you care about `GiftLineItemCartDiscountValue` when you don't sync it anyway?

File: src/test/java/com/commercetools/sync/cartdiscounts/utils/CartDiscountReferenceReplacementUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+class CartDiscountReferenceReplacementUtilsTest {
+
+    @Test
+    void replaceCartDiscountsReferenceIdsWithKeys_WithExpandedReferences_ShouldReturnReferencesWithReplacedKeys() {
+        final Type mockCustomType = getTypeMock(UUID.randomUUID().toString(), "customTypeKey");
+
+        final List<CartDiscount> mockCartDiscounts = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            final CartDiscount mockCartDiscount = mock(CartDiscount.class);
+            //Mock cartDiscounts custom fields with non-expanded type references.
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/cartdiscounts/utils/CartDiscountReferenceReplacementUtilsTest.java]: expanded or non-expended?

**Comment by heshamMassoud** [comment]
> Is there an integration test that test the whole flow from reference resolution to sync?

@lojzatran yes, in both:

`externalsource/cartdiscounts/CartDiscountSyncIT.java`
`ctpprojectsource/cartdiscounts/CartDiscountSyncIT.java`

They test syncing cart discounts with references (i.e. reference resolution is also being executed within)

File: src/test/java/com/commercetools/sync/cartdiscounts/utils/CartDiscountReferenceReplacementUtilsTest.java (hunk 3 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/cartdiscounts/utils/CartDiscountReferenceReplacementUtilsTest.java]: 8c44378

File: src/main/java/com/commercetools/sync/cartdiscounts/helpers/CartDiscountReferenceResolver.java (hunk 2 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/cartdiscounts/helpers/CartDiscountReferenceResolver.java]: thanks this is an outdated javadoc. Updated now: 
a7aebd6

**Pull request merged**